en/5-network/README.md

```diff
@@ -1,8 +1,497 @@
+```
+	//check if we have a good version
+	start := strings.Index(userAgent, ":")
+	end := strings.Index(userAgent[start:], "/")
+	if start < 0 && end < 0 {
+		panic(errors.New(fmt.Sprintf("cannot parse version in %s", userAgent)))
+	}
+	semVer := userAgent[start+1:start+end]
+	fmt.Printf("parsed semver: %v\n", semVer)
+	v1, err := version.NewVersion(semVer)
+	min, err := version.NewVersion("2.10.1")
+	if v1.LessThan(min) {
+		panic(errors.New(fmt.Sprintf("%s is less than %s", v1, min)))
+	}
+```
+
+Since we sent a version packet, we need to get the verack and we also need to send a verack, as we received the version as well:
+
+```
+	////////// got version, send ack
+	packetVerack := encodeHeader("verack", []byte{})
+	n, err = remote.Write(packetVerack);
+	if err != nil {
+		panic(err)
+	}
+
+	///////// wait for verack confirmation
+	read = make([]byte, 24)
+	n, err = io.ReadFull(remote, read)
+	plen, rcvChecksum = decodeHeader(read)
+	fmt.Printf("read verack array: %v, %d\n", read, plen)
+	if rcvChecksum != 3806393949 {
+		panic(errors.New("checksum mismatch in verack!"))
+	}
+```
+After the version/verack, we can now send the ping. Without the versions, no commands can be sent
+```
+	/////// send ping
+	packet2 := encodeHeader("ping", encodePing())
+	n, err = remote.Write(packet2)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("wrote ping: %v, %d\n", packet2, n)
+```
+After sending ping, we can expect a pong message.
+```
+	//////// receive pong
+	read = make([]byte, 36)
+	_, err = io.ReadFull(remote, read)
+	_, rcvChecksum = decodeHeader(read)
+	decodePing(read[24 : 24+12])
+	fmt.Printf("read array: %v\n", read)
+
+	tmp = sha256.Sum256(read[24 : 24+12])
+	hash = sha256.Sum256(tmp[:])
+	checksum = binary.LittleEndian.Uint32(hash[0:4])
+
+	if rcvChecksum != checksum {
+		panic(errors.New("checksum mismatch in pong!"))
+	}
+
+	remote.Close()
+}
+```
+---
+**Exercise 4**:
+Merge your encoder/decoder code with the main function from the template and run it on the mainnet. What can you see?
+---
+
+After successfully sending a NEO ping and receive a pong back, we can send `getaddr` and receive further nodes in the NEO networks. For that you can use the connection you have already established. Finding other peers is crucial for P2P and distributed systems, as peers may go offline at any time, and having other nodes to connect to is vital.
+
+---
+**Exercise 5**:
+After sending the ping message, send a `getaddr` message. Analyze the output.
+
+Hint:
+The output contains a list of IP addresses (16 bytes IPv6/4, 2 bytes port) that are encoded big-endian. The data also contains a service flag (set to 1) and a timestamp.
+---
+
```

Comment:
The paper is pretty good, and expect for better format.   @steven1227 @tbocek

Reply:
Great!

What do you mean by better format?

Reply:
I mean, could keep the format of `Exercise 4` and `Exercise 5`  the same with `Exercise1,2,3`, but If you want to emphasize them, you can keep the original format.

Reply:
![image](https://user-images.githubusercontent.com/43396869/58008788-b7d67a80-7b1f-11e9-91d4-0703dc6794c2.png)

Reply:
Good point! My markdown editor showed this a bit differently. I did not notice that. Its now fixed, and the layout should look the same in all exercises.